packages/lending/src/adapter.ts

```diff
@@ -185,6 +185,299 @@ export class Adapter extends common.Web3Toolkit {
     return this.protocolMap[id];
   }
 
+  async openByCollateral(
+    account: string,
+    portfolio: Portfolio,
+    zapToken: common.Token,
+    zapAmount: string,
+    collateralToken: common.Token,
+    collateralAmount: string,
+    debtToken: common.Token,
+    slippage = defaultSlippage
+  ): Promise<OperationOutput> {
+    const output: OperationOutput = {
+      destAmount: '0',
+      afterPortfolio: portfolio.clone(),
+      logics: [],
+    };
+    try {
+      // 1. ---------- validate params ----------
+      if (collateralAmount === '0') throw new OperationError('collateralAmount', 'ZERO_AMOUNT');
+
+      const srcCollateral = portfolio.findSupply(collateralToken);
+      if (!srcCollateral) throw new OperationError('collateralToken', 'UNSUPPORTED_TOKEN');
+
+      const destBorrow = portfolio.findBorrow(debtToken);
+      if (!destBorrow) throw new OperationError('debtToken', 'UNSUPPORTED_TOKEN');
+
+      const supplyAmountDelta = new common.TokenAmount(collateralToken.wrapped, collateralAmount).sub(
+        srcCollateral.balance
+      );
+      if (!srcCollateral.validateSupplyCap(supplyAmountDelta.amount)) {
+        throw new OperationError('collateralAmount', 'SUPPLY_CAP_EXCEEDED');
+      }
```

Comment:
Since this part involves validating parameters, should we also check if the amount is positive and throw error if it does?
```
if (Number(collateralAmount) < 0) throw new OperationError('collateralAmount', 'INVALID_AMOUNT');
```